### PR TITLE
Implement IAM

### PR DESCRIPTION
**Pull request by MykolaSolyanko**
Implement IAM
Add mTLS
Split vchan on to public and private

### CODE REVIEW

File: vchanmanager/vchanmanager.go [hunk 1]
```diff
@@ -141,15 +177,169 @@ func (v *VChanManager) Close() {
...
+	v.sendChan <- message{
+		msgSource: iam,
+		data:      data,
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+
+	case data = <-v.recvIAMChan:
```
Comment (al1img) [inline vchanmanager/vchanmanager.go:200]: We have `aos_common/utils/syncstream` specially designed for this case. See CM smcontroller, smhandler implementation for usage details.

File: vchanmanager/vchanmanager.go [hunk 2]
```diff
@@ -59,19 +70,40 @@ type Unpacker interface {
...
 type VChanItf interface {
 	Connect(ctx context.Context) error
-	ReadMessage() ([]byte, error)
-	WriteMessage(data []byte) error
+	ReadMessage() (message, error)
+	WriteMessage(msg message) error
 	Disconnect() error
 }
 
+type CertHandler interface {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager.go]: Done

[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager.go]: Done

File: vchanmanager/vchanmanager.go (hunk 1 above)
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager.go]: Done

File: vchanmanager/vchanmanager.go [hunk 3]
```diff
@@ -58,22 +58,21 @@ type Unpacker interface {
...
+	ReadMessage() ([]byte, error)
+	WriteMessage(data []byte) error
+	Disconnect() error
 }
 
 // Vchan vchan instance.
 type VChanManager struct {
 	sync.Mutex
 
-	vchanReader     VChanItf
```
Comment (al1img) [inline vchanmanager/vchanmanager.go:72]: Looks like mutex not used.

File: main.go [hunk 4]
```diff
@@ -125,10 +128,37 @@ func main() {
 		return
 	}
 
+	cryptoContext, err := cryptutils.NewCryptoContext(config.CACert)
+	if err != nil {
+		log.Fatalf("Can't create crypto context: %s", err) //nolint:gocritic
```
Comment (al1img) [inline main.go]: %v

File: main.go [hunk 5]
```diff
@@ -125,10 +128,37 @@ func main() {
...
+
+		return
+	}
+	defer cryptoContext.Close()
+
+	iam, err := iamclient.New(config, cryptoContext, false)
+	if err != nil {
+		log.Fatalf("Can't create iam client: %s", err)
```
Comment (al1img) [inline main.go]: ditto

File: vchanmanager/vchanmanager.go [hunk 6]
```diff
@@ -138,18 +163,163 @@ func (v *VChanManager) Close() {
...
+		log.Errorf("Failed to disconnect from vchan: %v", err)
+	}
 
-	close(v.recvChan)
+	close(v.recvSMChan)
 	close(v.sendChan)
 }
 
+// CreateKey creates key.
+func (v *VChanManager) CreateKey(ctx context.Context, req *pbIAM.CreateKeyRequest) (*pbIAM.CreateKeyResponse, error) {
```
Comment (al1img) [inline vchanmanager/vchanmanager.go:179]: All these API's should be part of IAM server package. `vchanmanager` should not be aware of what it sends/receives exactly and should not perform any conversion. `VChanManager` may have the interface as follows:
```go
type ReceiverItf interface {
	ReceiveMessage(requestID uint64, methodName string, data[]byte, err error)
}

func (v *VChanManager) SendMessage(source MessageSource,  methodName string, priv bool, data []byte) error
func (v *VChanManager) RegisterReceiver(source MessageSource, receiver ReceiverItf) error
```

File: vchanmanager/include/vchanapi.h [hunk 7]
```diff
@@ -25,6 +25,8 @@ struct VChanMessageHeader {
     uint32_t mSource;
     uint32_t mDataSize;
     uint64_t mRequestId;
+    int32_t  mErrno;
```
Comment (al1img) [inline vchanmanager/include/vchanapi.h:28]: Change tag in commit message: vendor for vendro folder

File: vchanmanager/vchan.go [hunk 8]
```diff
@@ -312,3 +353,34 @@ func prepareHeader(msg Message) []byte {
 
 	return (*[headerSize]byte)(unsafe.Pointer(&header))[:]
 }
+
+func convertAosErrorToError(err aosError) error {
```
Comment (al1img) [inline vchanmanager/vchan.go:357]: Use stringer interface: https://go.dev/tour/methods/17

File: vchanmanager/vchan.go [hunk 9]
```diff
@@ -312,3 +353,34 @@ func prepareHeader(msg Message) []byte {
...
+	case eWrongState:
+		return aoserrors.Errorf("wrong state")
+	case eInvalidChecksum:
+		return aoserrors.Errorf("invalid checksum")
+	default:
+		return aoserrors.Errorf("unknown error")
+	}
+}
+
+func convertErrnoToError(errno C.int) string {
```
Comment (al1img) [inline vchanmanager/vchan.go:384]: No need for wrap function.

File: vchanmanager/vchan.go [hunk 10]
```diff
@@ -125,45 +151,138 @@ func (v *VChan) Read() (data []byte, err error) {
...
+func (vc *connection) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+
 /***********************************************************************************************************************
  * Private
  **********************************************************************************************************************/
 
-func (v *VChan) readVchan(buffSize C.size_t) ([]byte, error) {
+func (v *VChan) init() (conn *connection, err error) {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply vchanmanager/vchan.go:270]: Done

File: vchanmanager/vchanmanager.go [hunk 11]
```diff
@@ -83,8 +82,8 @@ type VChanManager struct {
  **********************************************************************************************************************/
 
 var (
-	ErrContextCanceled = errors.New("operation canceled due to context cancellation")
-	ErrChecksumFailed  = errors.New("checksum validation failed")
+	ErrChecksumFailed = errors.New("checksum validation failed")
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager.go:105]: no, make it in lower case

File: vchanmanager/vchanmanager.go (hunk 3 above)
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager.go:72]: agree, removed

File: main.go (hunk 4 above)
Comment (MykolaSolyanko) [reply main.go]: Done

File: main.go (hunk 5 above)
Comment (MykolaSolyanko) [reply main.go]: Done

File: vchanmanager/vchanmanager.go (hunk 6 above)
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager.go:180]: Discuss to move this in separate ticket, https://jiraeu.epam.com/browse/EPMDAEPR-4359

File: vchanmanager/include/vchanapi.h (hunk 7 above)
Comment (MykolaSolyanko) [reply vchanmanager/include/vchanapi.h]: irrelevant

**Review by al1img — APPROVED** [review]

**Pull request merged**